magenta/models/polyphonic_rnn/polyphonic_rnn_lib.py

```diff
@@ -0,0 +1,1377 @@
+# Copyright 2016 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utilities for polyphonic RNN."""
+
+from collections import OrderedDict
+import copy
+import os
+import re
+import uuid
+
+# internal imports
+
+import numpy as np
+from scipy import linalg
+import tensorflow as tf
+
+import magenta.music as mm
+from magenta.protobuf import music_pb2
+
+
+##
+# begin metautils
+##
+
+
+def shape(x):
+  # Get shape of Variable through hacky hacky string parsing
+  shape_tup = repr(tf.Print(x, [tf.shape(x)])).split('shape=')[1]
+  shape_tup = shape_tup.split('dtype=')[0][:-1]
+  shape_tup = shape_tup.split(',')
+  shape_tup = [re.sub('[^0-9?]', '', s) for s in shape_tup]
+  # remove empty '' dims in 1D case
+  shape_tup = tuple([int(s) if s != '?' else -1
+                     for s in shape_tup if len(s) >= 1])
+  if sum([1 for s in shape_tup if s < 1]) > 1:
+    raise ValueError('too many ? dims')
+  return shape_tup
+
+
+def ndim(x):
+  return len(shape(x))
+
+
+# TODO: How can I correct this for shared / tied weights?
+def print_network(params):
+  n_params = sum([np.prod(shape(p)) for p in params])
+  tf.logging.info('Total number of parameters: %fM' % (n_params / float(1E6)))
+
+
+def dot(a, b):
+  # Generalized dot for nd sequences, assumes last axis is projection
+  # b must be rank 2
+  # rediculously hacky string parsing... wowie
+  a_tup = shape(a)
```

Comment:
I do:
```Python
def int_shape(x):
  """Returns the shape of a tensor as a tuple of integers or None entries.

  Adapted from keras.backend.tensorflow_backend.

  Args:
    x: A TensorFlow tensor object.

  Returns:
    A tuple of integers or None (for unknown dimensions) corresponding to the
    shape of `x`.
  """
  shape = x.get_shape()
  return tuple([i.__int__() for i in shape])


def ndim(x):
  """Returns the number of dimensions of the TensorFlow variable `x`.

  Args:
    x: A TensorFlow tensor object.

  Returns:
    An integer denoting the number of dimensions in `x`.
  """
  return x.get_shape().ndims


def dot(x, y):
  """Computes the dot product of two N-dimensional tensors.

  Uses the same conventions as np.dot, allowing for variable-length dimensions.
  Note that the last dimension of x and the second-to-last dimension of y must
  be equal and known ahead of time.

  Args:
    x: TensorFlow tensor object.
    y: TensorFlow tensor object.

  Returns:
    A TensorFlow tensor which is the dot product of `x` and `y`.
  """
  # Flatten all dimensions of x up to the last
  xt = tf.reshape(x, [-1, int_shape(x)[-1]])
  # Place the second-to-last dimension of y first (this trick from keras)
  y_permute_dim = list(range(ndim(y)))
  y_permute_dim = [y_permute_dim.pop(-2)] + y_permute_dim
  # and flatten all dimensions after the first (which was originally the
  # second-to-last dimension)
  yt = tf.reshape(tf.transpose(y, perm=y_permute_dim), [int_shape(y)[-2], -1])
  # Compute dot product of flattened matrices
  xy = tf.matmul(xt, yt)
  # Reshape to (x_1, x_2, ..., x_{N - 1}, y_1, y_2, ..., y_{M - 3}, y_{M - 1})
  # Using tf.reshape and tf.pack and tf.shape allows for variable dimensions
  xy = tf.reshape(xy, tf.pack([tf.shape(x)[n] for n in range(ndim(x) - 1)] +
                              [tf.shape(y)[n] for n in range(ndim(y) - 2)] +
                              [tf.shape(y)[ndim(y) - 1]]))
```

I think it may also be useful to call `set_shape` after the `reshape` because `reshape` makes all of the dimensions unknown even when they are known.